virtual_rainforest/models/hydrology/hydrology_model.py

```diff
@@ -702,3 +691,113 @@ def calculate_soil_evaporation(
+
+    Args:
+        neighbours: list of neighbour IDs
+        elevation: elevation, [m]
+
+    Returns:
+        list of lowest neighbour IDs
+    """
+    lowest_neighbour = []
+    for cell_id, neighbors_id in enumerate(neighbours):
+        downstream_id_loc = np.argmax(elevation[cell_id] - elevation[neighbors_id])
+        lowest_neighbour.append(neighbors_id[downstream_id_loc])
+
+    return lowest_neighbour
+
+
+def find_upstream_cells(lowest_neighbour: list) -> list:
+    """Find all upstream cell IDs for all grid cells.
+
+    This function identifies all cell IDs that are upstream of each grid cell. This can
+    be used to calculate the water flow that goes though a grid cell.
+
+    Args:
+        lowest_neighbour: list of lowest neighbour cell_ids
+
+    Returns:
+        list of all upstream IDs for each grid cell
+    """
+    upstream_ids: list = [[] for i in range(len(lowest_neighbour))]
+
+    for down_s, up_s in enumerate(lowest_neighbour):
+        upstream_ids[up_s].append(down_s)
+
+    return upstream_ids
+
+
+def accumulate_surface_runoff(
+    drainage_map: dict,
+    surface_runoff: np.ndarray,
+    accumulated_runoff: np.ndarray,
+) -> np.ndarray:
+    """Calculate accumulated surface runoff for each grid cell.
+
+    This function takes the accumulated surface runoff from the previous timestep and
+    adds all surface runoff of the current time step from upstream cell IDs.
+
+    The function currently raises a `ValueError` if accumulated runoff is negative.
+
+    Args:
+        drainage_map: dict of all upstream IDs for each grid cell
+        surface_runoff: surface runoff of the current time step, [mm]
+        accumulated_runoff: accumulated surface runoff from previous time step, [mm]
+
+    Returns:
+        accumulated surface runoff, [mm]
+    """
+
+    upstream_id_list = [[i for i in drainage_map[x]] for x in drainage_map.keys()]
+
+    for cell_id, upstream_id in enumerate(upstream_id_list):
+        accumulated_runoff[cell_id] += np.sum(surface_runoff[upstream_id])
+
+    for num in accumulated_runoff:
+        if num < 0:
+            to_raise = ValueError(
+                "The accumulated surface runoff should not be negative!"
+            )
+            LOGGER.error(to_raise)
+            raise to_raise
+
+    return accumulated_runoff
+
+
+# TODO move this to core.grid once we decided on common use
+def calculate_drainage_map(grid: Grid, elevation: np.ndarray) -> dict[int, tuple[int]]:
+    """Calculate drainage map based on digital elevation model.
+
+    This function finds the lowest neighbour for eacj grid cell, identifies all upstream
+    IDs and creates a dictionnary that provides all upstream cell IDs for each grid cell
+    . This function currently supports only square grids.
+
+    Args:
+        Grid: grid object
+        elevation: elevation, [m]
+
+    Returns:
+        tuple of cell IDs and their upstream neighbours
```

Comment:
I don't think this is what's being returned... right?

Reply:
yes, you're correct. I was looking at the tuple in the dict